spec/integration/indirector/report/yaml.rb

```diff
@@ -0,0 +1,48 @@
+require 'spec_helper'
+
+require 'puppet/transaction/report'
+require 'puppet/indirector/report/yaml'
+require 'puppet/indirector/indirector_testing/report/yaml'
+
+describe Puppet::Transaction::Report::Yaml do
+  describe '#save' do
+    let(:model) { Puppet::IndirectorTesting }
+    let(:indirection) { model.indirection }
+    let(:instance) { model.new('report') }
+    let(:request)  { indirection.request(:report, 'example', instance) }
+    let(:file) { subject.path(request.key) }
+
+    it 'saves the instance of the report as YAML to disk' do
+      subject.save(request)
+      content = File.read(file)
+      expect(content).to match(/value\s*:\s*report/)
+    end
+
+    it 'allows mode overwrite' do
+      Puppet.settings.setting(:lastrunreport).mode = '0644'
+      subject.save(request)
+
+      if Puppet::Util::Platform.windows?
+        require 'puppet/util/windows/security'
+        mode = Puppet::Util::Windows::Security.get_mode(file)
+      else
+        mode = Puppet::FileSystem.stat(file).mode
+      end
+
+      expect(mode & 07777).to eq(0644)
+    end
+
+    it 'raises error Puppet::DevError if mode is invalid' do
+      allow(Puppet.settings.setting(:lastrunreport)).to receive(:mode).and_return('9999')
+      expect{ subject.save(request) }.to \
+        raise_error(Puppet::DevError, "replace_file mode: 9999 is invalid")
+    end
+
+    it 'logs error when cannot save report' do
+      allow(Puppet::FileSystem).to receive(:replace_file).and_raise(TypeError.new('detail'))
```

Comment:
I try to avoid stubbing in an integration test unless it's unavoidable. For the case where the report can't be serialized to yaml, try setting something on the report that can't be serialized:

```
irb(main):005:0> report = Puppet::Transaction::Report.new
irb(main):006:0> report.configuration_version = Class.new
irb(main):007:0> YAML.dump(report)
Traceback (most recent call last):
...
TypeError (can't dump anonymous class: #<Class:0x00007f963932d490>)
```

Also would be good test what `save` raises if `replace_file` fails. For example, create a directory and then try to replace_file with the same path:

```
path = File.join(tmpdir('report'), report.name)
FileUtil.mkdir_p(path)
expect {
  Puppet::Transaction::Report.indirection.save(report)
}.to raise_error(...)
```

I'm not sure what `save` raises something like `Errno::EISDIR` or if that is wrapped in a `Puppet::Error`?

Reply:
It raises `Errno::EISDIR`. Should we also handle this?